### DISCUSSION TITLE
Request for Guidance on Setting Up Development Environment

### DISCUSSION THREAD

**Discussion by mahendrasinghbora** [hystax/optscale#305] [Q&A] [answered]
Hi,

I've noticed that there isn't any documentation on how to set up optscale for development. Could anyone provide advice or instructions on how to properly configure the project for a development environment?

Thanks in advance for your help!

**Comment by dcwire** [comment, marked as answer]
Hi, so the way I have set up a dev environment for myself is (atleast for the ngui part) by making use of `mirrord`. Mirrord creates a new pod in the kubernetes cluster that replicates the traffic going to a pod that you intend to develop (lets say the ngui pod) and sends that traffic to your locally running server. This way whatever changes you made to your backend will be reflected immediately how you would want it to while working locally without having to build the image everytime. It pulls the env variables from the pod and passes it to the local environment. 

Steps to reproduce: 
1. Install mirrord from the link https://mirrord.dev/docs/overview/quick-start/ 

2. Now the issue is that the node only allows 110 pods to be run and some pods would've already completed so you can delete them by using `kubectl delete jobs --field-selector status.successful=1` 

3. Go to your server folder in ngui, set it up how you would when running it locally. Can use this for reference https://github.com/hystax/optscale/issues/145 . Then run the command : 
`mirrord exec --target pod/pod_name --override-env-vars-exclude UI_BUILD_PATH pnpm start` 

We use the `--override-env-vars-exclude` flag to exclude env variables from the pod being loaded so we can make use of local env variables. The reason why I ignored the UI_BUILD_PATH was because inside the docker container the path to the build file was different and outside the container its different, so I wanted to make use of the local UI_BUILD_PATH to be used instead.
